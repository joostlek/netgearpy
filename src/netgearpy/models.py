"""NetgearPy models for device settings."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum, StrEnum
from typing import Any

from mashumaro import field_options
from mashumaro.mixins.dict import DataClassDictMixin

from netgearpy.helpers import statistics_to_data, time_to_minutes


@dataclass
class CurrentSettings(DataClassDictMixin):
    """Represents the current setting of a device."""

    firmware_version: str = field(metadata=field_options(alias="Firmware"))
    model: str = field(metadata=field_options(alias="Model"))
    login_method: int = field(metadata=field_options(alias="LoginMethod"))
    https_port: int = field(metadata=field_options(alias="SOAP_HTTPs_Port"))

    @classmethod
    def __pre_deserialize__(cls, d: dict[str, Any]) -> dict[str, Any]:
        """Pre deserialize hook."""
        return {
            **d,
            "LoginMethod": int(float(d["LoginMethod"])),
        }


class ConnectionType(StrEnum):
    """Enum for connection types."""

    WIRED = "wired"
    WIRELESS = "wireless"


@dataclass(kw_only=True)
class AttachedDevice:
    """Represents an attached device."""

    ip_address: str
    hostname: str | None
    mac_address: str
    connection_type: ConnectionType | None
    link_speed: int | None
    signal_strength: int | None
    blocked: bool | None

    @classmethod
    def from_string(cls, source_string: str) -> AttachedDevice:
        """Pre deserialize hook."""
        parts = source_string.split(";")
        ip_address = parts[1]
        hostname = None
        if parts[2] != "--":
            hostname = parts[2]
        mac_address = parts[3]
        connection_type = None
        link_speed = None
        signal_strength = None
        blocked = None
        if len(parts) >= 7:
            connection_type = ConnectionType(parts[4])
            if parts[5] != "":
                link_speed = int(parts[5])
            signal_strength = int(parts[6])
            if len(parts) >= 8:
                blocked = parts[7] != "Allow"
        return AttachedDevice(
            ip_address=ip_address,
            hostname=hostname,
            mac_address=mac_address,
            connection_type=connection_type,
            link_speed=link_speed,
            signal_strength=signal_strength,
            blocked=blocked,
        )


class DeviceMode(IntEnum):
    """Enum for device modes."""

    ROUTER = 0
    ACCESS_POINT = 1


@dataclass
class DeviceInfo(DataClassDictMixin):
    """Represents the device information."""

    model: str = field(metadata=field_options(alias="ModelName"))
    serial_number: str = field(metadata=field_options(alias="SerialNumber"))
    firmware_version: str = field(metadata=field_options(alias="Firmwareversion"))
    smart_agent_version: str = field(metadata=field_options(alias="SmartAgentversion"))
    firewall_version: str = field(metadata=field_options(alias="FirewallVersion"))
    vpn_version: str | None = field(metadata=field_options(alias="VPNVersion"))
    other_software_version: str = field(
        metadata=field_options(alias="OthersoftwareVersion")
    )
    hardware_version: str = field(metadata=field_options(alias="Hardwareversion"))
    other_hardware_version: str | None = field(
        metadata=field_options(alias="Otherhardwareversion")
    )
    device_name: str = field(metadata=field_options(alias="DeviceName"))
    device_mode: DeviceMode = field(metadata=field_options(alias="DeviceMode"))

    @classmethod
    def __pre_deserialize__(cls, d: dict[str, Any]) -> dict[str, Any]:
        """Pre deserialize hook."""
        if d["VPNVersion"] == "N/A":
            d["VPNVersion"] = None
        if d["Otherhardwareversion"] == "N/A":
            d["Otherhardwareversion"] = None
        return {**d, "DeviceMode": int(d["DeviceMode"])}


@dataclass
class TrafficMeterStatistics(DataClassDictMixin):
    """Represents the traffic meter statistics."""

    connection_time_today: int = field(
        metadata=field_options(alias="NewTodayConnectionTime")
    )
    upload_today: float = field(metadata=field_options(alias="NewTodayUpload"))
    download_today: float = field(metadata=field_options(alias="NewTodayDownload"))
    connection_time_yesterday: int = field(
        metadata=field_options(alias="NewYesterdayConnectionTime")
    )
    upload_yesterday: float = field(metadata=field_options(alias="NewYesterdayUpload"))
    download_yesterday: float = field(
        metadata=field_options(alias="NewYesterdayDownload")
    )
    connection_time_week: int = field(
        metadata=field_options(alias="NewWeekConnectionTime")
    )
    total_upload_week: float
    average_upload_week: float
    total_download_week: float
    average_download_week: float
    connection_time_month: int = field(
        metadata=field_options(alias="NewMonthConnectionTime")
    )
    total_upload_month: float
    average_upload_month: float
    total_download_month: float
    average_download_month: float
    connection_time_last_month: int = field(
        metadata=field_options(alias="NewLastMonthConnectionTime")
    )
    total_upload_last_month: float
    average_upload_last_month: float
    total_download_last_month: float
    average_download_last_month: float

    @classmethod
    def __pre_deserialize__(cls, d: dict[str, Any]) -> dict[str, Any]:
        """Pre deserialize hook."""
        total_upload_week, average_upload_week = statistics_to_data(d["NewWeekUpload"])
        total_download_week, average_download_week = statistics_to_data(
            d["NewWeekDownload"]
        )
        total_upload_month, average_upload_month = statistics_to_data(
            d["NewMonthUpload"]
        )
        total_download_month, average_download_month = statistics_to_data(
            d["NewMonthDownload"]
        )
        total_upload_last_month, average_upload_last_month = statistics_to_data(
            d["NewLastMonthUpload"]
        )
        total_download_last_month, average_download_last_month = statistics_to_data(
            d["NewLastMonthDownload"]
        )
        return {
            **d,
            "NewTodayConnectionTime": time_to_minutes(d["NewTodayConnectionTime"]),
            "NewYesterdayConnectionTime": time_to_minutes(
                d["NewYesterdayConnectionTime"]
            ),
            "NewWeekConnectionTime": time_to_minutes(d["NewWeekConnectionTime"]),
            "total_upload_week": total_upload_week,
            "average_upload_week": average_upload_week,
            "total_download_week": total_download_week,
            "average_download_week": average_download_week,
            "NewMonthConnectionTime": time_to_minutes(d["NewMonthConnectionTime"]),
            "total_upload_month": total_upload_month,
            "average_upload_month": average_upload_month,
            "total_download_month": total_download_month,
            "average_download_month": average_download_month,
            "NewLastMonthConnectionTime": time_to_minutes(
                d["NewLastMonthConnectionTime"]
            ),
            "total_upload_last_month": total_upload_last_month,
            "average_upload_last_month": average_upload_last_month,
            "total_download_last_month": total_download_last_month,
            "average_download_last_month": average_download_last_month,
        }
