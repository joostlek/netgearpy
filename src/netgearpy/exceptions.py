"""Asynchronous Python client for Netgear routers."""


class NetgearError(Exception):
    """Generic exception."""
