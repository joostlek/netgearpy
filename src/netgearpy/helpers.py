"""Helper functions for NetgearPy."""


def time_to_minutes(source: str) -> int:
    """Convert a time string to minutes."""
    parts = source.split(":")
    if parts[0] == "--" and parts[1] == "--":
        return 0
    hours = int(parts[0])
    minutes = int(parts[1])
    return hours * 60 + minutes


def statistics_to_data(source: str) -> tuple[float, float]:
    """Convert a statistics string to a dictionary."""
    parts = source.split("/")
    return float(parts[0]), float(parts[1])
