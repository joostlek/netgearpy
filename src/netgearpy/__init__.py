"""Asynchronous Python client for Netgear routers."""

from .models import CurrentSettings
from .netgear import NetgearClient

__all__ = [
    "CurrentSettings",
    "NetgearClient",
]
