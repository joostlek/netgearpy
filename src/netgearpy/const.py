"""Constants for NetgearPy."""

from netgearpy.models import Service

ENVELOPE = """<!--?xml version="1.0" encoding= "UTF-8" ?-->
<v:Envelope
xmlns:v="http://schemas.xmlsoap.org/soap/envelope/">
<v:Header>
<SessionID>A7D88AE69687E58D9A00</SessionID>
</v:Header>
<v:Body>
{}
</v:Body>
</v:Envelope>"""

EMPTY_BODY = """<M1:{} xmlns:M1="urn:NETGEAR-ROUTER:service:{}:1" />"""

LOGIN_BODY = """<M1:SOAPLogin xmlns:M1="urn:NETGEAR-ROUTER:service:DeviceConfig:1">
<Username>{}</Username><Password>{}</Password></M1:SOAPLogin>"""


def _get_empty_body(service: Service, object_type: str) -> str:
    """Get an empty body for a SOAP request."""
    return EMPTY_BODY.format(object_type, service)


GET_ATTACHED_DEVICES_BODY = _get_empty_body(Service.DEVICE_INFO, "GetAttachDevice")

GET_INFO_BODY = _get_empty_body(Service.DEVICE_INFO, "GetInfo")

GET_TRAFFIC_METER_STATISTICS_BODY = _get_empty_body(
    Service.DEVICE_CONFIG, "GetTrafficMeterStatistics"
)

IS_PARENTAL_CONTROL_ENABLED_BODY = """<GetEnableStatus></GetEnableStatus>"""

GET_ETHERNET_LINK_STATUS_BODY = """<M1:GetEthernetLinkStatus xsi:nil="true" />"""
