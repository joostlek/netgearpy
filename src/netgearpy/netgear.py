"""Asynchronous Python client for Netgear routers."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from importlib import metadata
from typing import Any, Self
import xml.etree.ElementTree as ET

from aiohttp import ClientSession
from aiohttp.hdrs import METH_GET, METH_POST
from yarl import URL

from netgearpy.const import (
    ENVELOPE,
    GET_ATTACHED_DEVICES_BODY,
    GET_BLOCK_DEVICE_ENABLE_BODY,
    GET_ETHERNET_LINK_STATUS_BODY,
    GET_INFO_BODY,
    GET_TRAFFIC_METER_STATISTICS_BODY,
    IS_PARENTAL_CONTROL_ENABLED_BODY,
    LOGIN_BODY,
)
from netgearpy.models import (
    AttachedDevice,
    CurrentSettings,
    DeviceConfigAction,
    DeviceInfo,
    DeviceInfoAction,
    ParentalControlAction,
    Service,
    TrafficMeterStatistics,
    WanEthernetLinkConfigAction,
)

VERSION = metadata.version(__package__)


@dataclass
class NetgearClient:
    """Main class for handling connections with a Netgear router."""

    host: str
    session: ClientSession | None = None
    request_timeout: int = 10
    _close_session: bool = False
    _soap_port: int | None = None
    _login_method: int | None = None

    async def _request(
        self,
        url: str,
        *,
        method: str = METH_GET,
        data: str | None = None,
        headers: dict[str, Any] | None = None,
    ) -> str:
        """Handle a request to a Netgear router."""
        base_headers = {
            "User-Agent": f"Netgearpy/{VERSION}",
            "Accept": "application/xml",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        if headers is None:
            headers = {}

        async with asyncio.timeout(self.request_timeout):
            response = await self.session.request(
                method,
                url,
                headers=base_headers | headers,
                data=data,
            )

        return await response.text()

    async def _get(self, uri: str) -> str:
        """Handle a GET request to Netgear."""
        return await self._request(uri, method=METH_GET)

    async def get_current_setting(self) -> CurrentSettings:
        """Get the current settings from the Netgear router."""
        url = URL.build(scheme="http", host=self.host, path="/currentsetting.htm")
        response = await self._get(str(url))
        res: dict[str, Any] = {}
        for entry in response.split("\n"):
            if entry:
                split_entry = entry.split("=")
                res[split_entry[0]] = split_entry[1]
        result = CurrentSettings.from_dict(res)
        self._soap_port = result.https_port
        self._login_method = result.login_method
        return result

    async def _post_xml(
        self, service: str, action: str, body: str
    ) -> dict[str, str | None]:
        """Post XML data to the Netgear router."""
        if self._soap_port is None or self._login_method is None:
            await self.get_current_setting()
        https = self._soap_port in [443, 5555]
        url = URL.build(
            scheme="https" if https else "http",
            host=self.host,
            path="/soap/server_sa/",
            port=self._soap_port,
        )
        headers = {
            "SOAPAction": f"urn:NETGEAR-ROUTER:service:{service}:1#{action}",
        }
        body_str = ENVELOPE.format(body)
        response = await self._request(
            str(url), method=METH_POST, data=body_str, headers=headers
        )
        response_dict = {}
        root = ET.fromstring(response)  # noqa: S314
        for item in root.iter():
            if item.tag != "{http://schemas.xmlsoap.org/soap/envelope/}Body":
                continue
            for child in item:
                for subchild in child:
                    response_dict[subchild.tag] = subchild.text
        return response_dict

    async def login(self, username: str, password: str) -> None:
        """Login to the Netgear router."""
        await self._post_xml(
            Service.DEVICE_CONFIG,
            DeviceConfigAction.LOGIN,
            LOGIN_BODY.format(username, password),
        )

    async def get_attached_devices(self) -> list[AttachedDevice]:
        """Get the attached devices from the Netgear router."""
        response = await self._post_xml(
            Service.DEVICE_INFO,
            DeviceInfoAction.GET_ATTACHED_DEVICES,
            GET_ATTACHED_DEVICES_BODY,
        )
        device_string = response["NewAttachDevice"]
        if not device_string:
            return []
        return [
            AttachedDevice.from_string(entry) for entry in device_string.split("@")[1::]
        ]

    async def is_parental_control_enabled(self) -> bool:
        """Check if parental control is enabled."""
        response = await self._post_xml(
            Service.PARENTAL_CONTROL,
            ParentalControlAction.GET_ENABLE_STATUS,
            IS_PARENTAL_CONTROL_ENABLED_BODY,
        )
        return response.get("ParentalControl") == "1"

    async def is_block_device_enabled(self) -> bool:
        """Check if block device is enabled."""
        response = await self._post_xml(
            Service.DEVICE_CONFIG,
            DeviceConfigAction.GET_BLOCK_DEVICE_ENABLE_STATUS,
            GET_BLOCK_DEVICE_ENABLE_BODY,
        )
        return response.get("ParentalControl") == "1"

    async def get_device_info(self) -> DeviceInfo:
        """Get device information from the Netgear router."""
        response = await self._post_xml(
            Service.DEVICE_INFO, DeviceInfoAction.GET_INFO, GET_INFO_BODY
        )
        return DeviceInfo.from_dict(response)

    async def get_traffic_meter_statistics(self) -> TrafficMeterStatistics:
        """Get traffic meter statistics from the Netgear router."""
        response = await self._post_xml(
            Service.DEVICE_CONFIG,
            DeviceConfigAction.GET_TRAFFIC_METER_STATISTICS,
            GET_TRAFFIC_METER_STATISTICS_BODY,
        )
        return TrafficMeterStatistics.from_dict(response)

    async def get_ethernet_link_status(self) -> str:
        """Get the ethernet link status from the Netgear router."""
        response = await self._post_xml(
            Service.WAN_ETHERNET_LINK_CONFIG,
            WanEthernetLinkConfigAction.GET_ETHERNET_LINK_STATUS,
            GET_ETHERNET_LINK_STATUS_BODY,
        )
        assert response["NewEthernetLinkStatus"] is not None  # noqa: S101
        return response["NewEthernetLinkStatus"]

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The NetgearClient object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
