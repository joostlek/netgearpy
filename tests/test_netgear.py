"""Asynchronous Python client for Netgear routers."""

from __future__ import annotations

from typing import TYPE_CHECKING

from aiohttp.hdrs import METH_POST
from aioresponses import aioresponses

from . import load_fixture
from .const import HEADERS, MOCK_URL

if TYPE_CHECKING:
    from netgearpy.netgear import NetgearClient
    from syrupy import SnapshotAssertion


def add_soap_response_fixture(
    responses: aioresponses,
    fixture_name: str,
    status: int = 200,
) -> None:
    """Add a SOAP response fixture."""
    responses.post(
        f"{MOCK_URL}/soap/server_sa/",
        status=status,
        body=load_fixture(fixture_name),
    )


def check_soap_called(
    responses: aioresponses,
    soap_action: str,
    body: str,
) -> None:
    """Check if the SOAP request was called with the expected parameters."""
    responses.assert_called_with(
        f"{MOCK_URL}/soap/server_sa/",
        METH_POST,
        headers=HEADERS | {"SOAPAction": soap_action},
        data=f"""<!--?xml version="1.0" encoding= "UTF-8" ?-->\n<v:Envelope\nxmlns:v="http://schemas.xmlsoap.org/soap/envelope/">\n<v:Header>\n<SessionID>A7D88AE69687E58D9A00</SessionID>\n</v:Header>\n<v:Body>\n{body}\n</v:Body>\n</v:Envelope>""",
    )


async def test_get_current_setting(
    responses: aioresponses, snapshot: SnapshotAssertion, netgear_client: NetgearClient
) -> None:
    """Test get current router setting."""
    responses.get(
        "http://192.168.0.1/currentsetting.htm",
        status=200,
        body=load_fixture("current_setting.txt"),
    )
    assert await netgear_client.get_current_setting() == snapshot


async def test_login(responses: aioresponses, netgear_client: NetgearClient) -> None:
    """Test login to the Netgear router."""
    responses.get(
        "http://192.168.0.1/currentsetting.htm",
        status=200,
        body=load_fixture("current_setting.txt"),
    )
    add_soap_response_fixture(responses, "login_response.xml")
    await netgear_client.login("username", "password")
    check_soap_called(
        responses,
        "urn:NETGEAR-ROUTER:service:DeviceConfig:1#SOAPLogin",
        """<M1:SOAPLogin xmlns:M1="urn:NETGEAR-ROUTER:service:DeviceConfig:1">
<Username>username</Username><Password>password</Password></M1:SOAPLogin>""",
    )


async def test_get_attached_devices(
    responses: aioresponses, snapshot: SnapshotAssertion, netgear_client: NetgearClient
) -> None:
    """Test get attached devices."""
    responses.get(
        "http://192.168.0.1/currentsetting.htm",
        status=200,
        body=load_fixture("current_setting.txt"),
    )
    add_soap_response_fixture(responses, "GetAttachDevice.xml")
    assert await netgear_client.get_attached_devices() == snapshot
    check_soap_called(
        responses,
        "urn:NETGEAR-ROUTER:service:DeviceInfo:1#GetAttachedDevices",
        """<M1:GetAttachedDevice xmlns:M1="urn:NETGEAR-ROUTER:service:DeviceInfo:1" />""",  # noqa: E501
    )
