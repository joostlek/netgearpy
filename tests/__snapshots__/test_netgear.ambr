# serializer version: 1
# name: test_get_attached_devices
  list([
    dict({
      'blocked': False,
      'connection_type': <ConnectionType.WIRELESS: 'wireless'>,
      'hostname': 'motorola-edge-50-fusion',
      'ip_address': '192.168.0.2',
      'link_speed': 192,
      'mac_address': 'D6:2E:BD:9B:0A:63',
      'signal_strength': 100,
    }),
    dict({
      'blocked': None,
      'connection_type': None,
      'hostname': 'motorola-edge-50-fusion',
      'ip_address': '192.168.0.2',
      'link_speed': None,
      'mac_address': 'D6:2E:BD:9B:0A:63',
      'signal_strength': None,
    }),
    dict({
      'blocked': None,
      'connection_type': <ConnectionType.WIRELESS: 'wireless'>,
      'hostname': 'motorola-edge-50-fusion',
      'ip_address': '192.168.0.2',
      'link_speed': 192,
      'mac_address': 'D6:2E:BD:9B:0A:63',
      'signal_strength': 100,
    }),
  ])
# ---
# name: test_get_current_setting
  dict({
    'firmware_version': 'V1.0.12.216_10.2.122',
    'https_port': 5555,
    'login_method': 2,
    'model': 'R7000',
  })
# ---
